### PR TITLE
[nikohomecontrol] Use DS annotations, get broadcast address from configuration.

### LINKED ISSUE #3054: Several addons are not doing broadcast packets correctly
[Several addons](https://github.com/openhab/openhab2-addons/search?utf8=%E2%9C%93&q=%22255.255.255.255%22&type=) are not doing discovery via broadcast packets correctly. They hardcode 255.255.255.255 which will not work on all networks. Since the method of getting the correct broadcast addresses is non-trivial to get right I'd propose putting it in a common location for use by addons needing to use broadcast packets.

Where would be the preferred location for this common code? Would it be preferred that the addon simply provides a packet contents and a callback function and the rest is handled by common code or would it be preferred that the common code just gets the broadcast addresses?

For reference, here is the code I came up with to find the addresses:

```java
    private List<InetAddress> getBroadcastAddresses() {
        ArrayList<InetAddress> addresses = new ArrayList<>();

        System.setProperty("java.net.preferIPv4Stack", "true");

        try {
            for (NetworkInterface iface : Collections.list(NetworkInterface.getNetworkInterfaces())) {
                try {
                    if (iface.isLoopback()) {
                        logger.debug("Found loopback iface {}", iface.getName());
                        continue;
                    }

                    logger.debug("Found non-loopback iface {} with {} addresses", iface.getName(),
                            iface.getInterfaceAddresses().size());

                    for (InterfaceAddress address : iface.getInterfaceAddresses()) {
                        InetAddress broadcastAddress = address.getBroadcast();
                        if (broadcastAddress != null) {
                            logger.debug("Found broadcast address {} for iface {}", address.getBroadcast().toString(),
                                    iface.getName());
                            addresses.add(address.getBroadcast());
                        }
                    }
                } catch (SocketException e) {
                    logger.debug("Got exception in getBroadcastAddress", e);
                }
            }
        } catch (SocketException e) {
            logger.debug("Got exception in getBroadcastAddress", e);
        }

        try {
            // Add the default broadcast address as a fallback.
            addresses.add(InetAddress.getByName("255.255.255.255"));
        } catch (UnknownHostException e) {
        }
        return addresses;
    }
```

### PR DESCRIPTION
**Pull request by mherwege**
Signed-off-by: Mark Herwege <mark.herwege@telenet.be>

Fixes #3054 
Requires smarthome [#4848](https://github.com/eclipse/smarthome/pull/4848)
Please do not merge before this is available to openHAB.

### CODE REVIEW

**Comment by martinvw** [comment]
@kaikreuzer does this first require a new release of ESH?

**Comment by kaikreuzer** [comment]
Yes, I plan this for next Monday.

**Comment by martinvw** [comment]
Great!

**Comment by martinvw** [comment]
@kaikreuzer where you able to do so?

**Comment by kaikreuzer** [comment]
@martinvw Yes, see https://community.openhab.org/t/first-major-update-in-2-3-0-snapshot/38801/.

File: addons/binding/org.openhab.binding.nikohomecontrol/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -1,27 +1,29 @@
...
+ org.eclipse.smarthome.config.discovery,
+ org.eclipse.smarthome.core.library.types,
+ org.eclipse.smarthome.core.net,
+ org.eclipse.smarthome.core.thing,
+ org.eclipse.smarthome.core.thing.binding,
+ org.eclipse.smarthome.core.types,
+ org.openhab.binding.nikohomecontrol,
+ org.openhab.binding.nikohomecontrol.handler,
+ org.osgi.framework,
+ org.osgi.service.component.annotations;version="1.3.0",
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.nikohomecontrol/META-INF/MANIFEST.MF]: please remove the version constraint and make the import optional

**Comment by martinvw** [comment]
Ah was that the same, yes I noticed that one 😉

@mherwege can you check what is wrong travis is not happy

**Comment by mherwege** [comment]
I struggle with the compile failure. It all works fine in my IDE.
Here is the error from Travis:
```
[ERROR] /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/discovery/NikoHomeControlBridgeDiscoveryService.java:[59] 
[ERROR] 	String broadcastAddr = networkAddressService.getConfiguredBroadcastAddress();
[ERROR] 	                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] The method getConfiguredBroadcastAddress() is undefined for the type NetworkAddressService
[ERROR] -> [Help 1]
```
getConfiguredBroadcastAddress() is the new method introduced in the NetworkAddressService interface in the ESH core package. Travis does not find it when compiling, although it should be in the current ESH master.
Any ideas?

File: addons/binding/org.openhab.binding.nikohomecontrol/META-INF/MANIFEST.MF (hunk 1 above)
Comment (mherwege) [reply addons/binding/org.openhab.binding.nikohomecontrol/META-INF/MANIFEST.MF]: Corrected.

**Comment by kaikreuzer** [comment]
> Travis does not find it when compiling, although it should be in the current ESH master.

It still seems to compile against an older ESH version. I'll today do another ESH stable and afterwards will try to figure out what's wrong here, if the problem remains.

**Comment by kaikreuzer** [comment]
@mherwege I managed to identify the issue and have a fix: https://github.com/openhab/openhab-core/pull/283.
The code of this PR is fine, I verified the build locally.

**Comment by Hilbrand** [comment]
@mherwege The `.gitignore` file in the OSGI-INF folder is empty, while it should contain `*.xml`. Without it, It results in unwanted untracked xml files.

**Comment by mherwege** [comment]
@Hilbrand. I created a new pull request [PR#3211](https://github.com/openhab/openhab2-addons/pull/3211) to correct this. Thanks for the message.

**Pull request merged**